### PR TITLE
✨ Get staging profile task

### LINKED ISSUE #60: ✨ Feature Request: GetStagingProfileTask
Is there any desired to incorporate the diagnostic `getStagingProfile` task from the `gradle-nexus-staging-plugin` to this project? I did a quick prototype and imagined it might look something like this

```kotlin
class GetStagingProfile @Inject constructor(
    objects: ObjectFactory,
    extension: NexusPublishExtension,
    repository: NexusRepository
) : AbstractNexusStagingRepositoryTask(objects, extension, repository) {

    @Optional
    @Input
    val packageGroup = objects.property<String>().apply {
        set(extension.packageGroup)
    }

    @TaskAction
    fun getStagingProfile() {
        val repository = repository.get()
        val serverUrl = repository.nexusUrl.get()
        val client = NexusClient(serverUrl, repository.username.orNull, repository.password.orNull, clientTimeout.orNull, connectTimeout.orNull)
        val stagingProfileId = client.findStagingProfileId(packageGroup = packageGroup.get()) ?: throw GradleException("Failed to find staging profile for package group: $packageGroup")
        logger.lifecycle("stagingProfileId '{}'", stagingProfileId)
    }
}
```

But I wasn't sure if this was left out unintentionally for strategic purposes - i'd be happy to contribute it with tests + update to the migration guide if this is something in line with your vision(s) for this project

As an aside, thank you for making this project! I stumbled upon it while publishing my very first artifact to maven central - this project made that a lot more approachable! ⬆️

### PR DESCRIPTION
**Pull request by ryandens**
Closes #60 

## Documentation Updates
I think we should add an update to the "Tasks" section of the migration guide to say something like 

```bash
./gradlew getStagingProfile
```

↓

```bash
./gradlew retrieveSonatypeStagingProfile
```

This reflects the main difference between the previous plugin project and this one: tasks are on a per `NexusRepsitory` basis, including the `RetrieveStagingProfile` task. 

## Summary of changes
First adds a new `AbstractNexusStagingRepositoryTask` subclass called `RetrieveStagingProfile`. we register a new `RetrieveStagingProfile` task for each `NexusRepository` registered in the plugin's extension. This enables users to run 
`./gradlew retrieve<NEXUS_REPOSITORY_NAME>StagingProfile` and see the staging profile ID for the repository in stdout. 

~This was initially as far as I was planning on going with this, but something @szpak mentioned in issue #60 caught my attention~
>  Or maybe there should be also an @Output field set to allow others to read it (but I'm not sure why anyone would like to consume that)

~After reading this, it struck me that we wouldn't want there to be inconsistencies between how  `GetStagingProfile` gets the staging profile and how `InitializeNexusStagingRepository` gets the staging profile. So, I went about a small refactor of the plugin to remove the responsibility of figuring out the staging profile id from `InitializeNexusStagingRepository` and instead make it a required input. Then, I configured `InitializeNexusStagingRepository` to depend on the `GetStagingProfile` which now has all the logic for getting the staging profile ID, outputting it to stdout, storing the ID in a file in the build directory, and managing the optimization of not doing the API lookup if the staging profile id is provided via configuration.~

~This not only has benefits for code re-use, but it also helps those who don't configure the stagingProfileId ever. Since we output the staging profile ID to an artifact, it can be cached and reused on subsequent builds which decreases the configuration burden on the user while still giving them access to this performance optimization.~

~I know changing `InitializeNexusStagingRepository` was not technically in the scope of what was discussed in #60 so if you would like me to remove it from the scope of this change I'd be happy to. But, it seemed to me like refactoring `InitializeNexusStagingRepository` was the responsible thing to do as I was introducing some logic duplication around how this plugin gets the staging profile id.~

Update on 2/27/21: After discussion, it was deemed that this refactor was problematic due to the conflicting nature of the different use cases of the `InitializeNexusStagingRepository` task and the diagnostic task to retrieve the staging profile

### Possible other tasks
Another thought that occurred to me was adding a task called `getStagingProfiles` or `getAllStagingProfiles` which is simply an alias for executing `get<NEXUS_REPOSITORY_NAME>StagingProfile` on each configured `NexusRepository`. I ultimately decided it was not worthwhile/might cause confusion but I'm willing to reconsider. 

### Testing gaps?
After writing up this pull request description, I realized my implementation for storing the staging repository ID was flawed, as I at first would overwrite the file storing the staging profile ID every time a new `GetStagingProfile` was run, which could lead to confusing results with multiple  `NexusRepository` destinations. The existing tests didn't catch it because there were no tests that published to two different repositories at the same time. I had to make some tweaks to the helper methods to accomplish this test, but I'm super open to feedback on how this was tested. Perhaps it would be better as a unit test than a compatTest?

### CODE REVIEW

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt
Comment (szpak) [inline src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt:280]: We don't enforce a declared order of tests, nor skill other tests after the first failure, but for logic order in operations performed in read, I would put it as the second test in the class (right after `must be applied to root project`).

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:71]: I like the idea of reusing the `getStagingProfile` task by the initialization task. However, I see a side effect which disturbs the silence of a regular Gradle tasks executions. On every publishing there will be a few lines displayed (task name, empty line and the message). Especially, having `stagingProfileId` explicitly configured it would be unexpected.

Possibly, we could detect if the task was explicitly called from the command line (e.g. using `gradle.startParameter.taskNames`) and only then display that message at the lifecycle level (otherwise on info). That might be not very elegant, but maybe @marcphilipp knows some Gradle trick to tweak it.

In addition, `Received staging profile id` is not fully true for the explicitly configured value (it was not received) and might be confusing for people migrating from GNSP.  Maybe there could be two variants of the output:
> Configured staging profile id: ...

and 
> Received staging profile id: ...

used depending on configuration. Nevertheless, I'm not fixed on that.

**Review by szpak — CHANGES_REQUESTED** [review]
Really informative PR description, thanks!

I went through the code and in general I like it. Chaining tasks makes sense, but there is one side effect related to logging. I put some remarks inline.

I also requested [e2e tests](https://github.com/gradle-nexus/publish-plugin/actions/runs/579625776) on your PR.

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt
Comment (ryandens) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt:212]: Ah, I definitely meant to come back here and clean this up prior to opening the PR, but I totally forgot! I tried to leverage the JUnit extension for providing a new instance of a WireMockServer, but it seems to not be supported. I think we have a couple of options
1. Decorate the WireMockServer with a `Closeable` implementation that calls the appropriate shutdown methods on the server
2. Write a small junit extension to manage the lifecycle of another WireMockServer. This is nice because it lets us re-use some of the existing components of the junit5 extension already in use by this project.



This is what the `Closeable` implementation/usgae would look like


```

    class CloseableWireMockServerContainer(val port: Int, private val server: WireMockServer = WireMockServer(port)) : Closeable,
        Supplier<WireMockServer> {
        override fun close() {
            server.stop()
        }

        override fun get(): WireMockServer {
            return server
        }
    }

@Test
    fun `publishes to two Nexus repositories`() {
      CloseableWireMockServerContainer(0). use {
      
      val otherServer = it.get()
      // omitted, rest of test

  }
}
```

Or, this is what the JUnit extension would look like 
```
/**
 * Composes a [WiremockResolver] and uses that by default. But, if a parameter is annotated
 * with [MethodScopeWiremockResolver] it creates a new instance of the [WiremockResolver] extension and
 * manages its lifecycle in the scope of that [ExtensionContext]
 */
class MethodScopeWiremockResolver(
    private val inner: WiremockResolver = WiremockResolver(),
    private val localResolvers: MutableMap<String, WiremockResolver> = mutableMapOf()
) : ParameterResolver by inner, AfterEachCallback {

    /**
     * Checks to see if a method the parameter is annotated with [MethodScopedWiremockServer]. If it is, we first verify
     */
    override fun resolveParameter(parameterContext: ParameterContext, extensionContext: ExtensionContext): Any {
        return if (parameterContext.isAnnotated(MethodScopedWiremockServer::class.java)) {
            if (!parameterContext.parameter.type.isAssignableFrom(WireMockServer::class.java)) {
                throw IllegalArgumentException("Annotated type must be a WireMockServer")
            }
            return localResolvers.getOrPut(extensionContext.uniqueId, { WiremockResolver() })
                .resolveParameter(parameterContext, extensionContext)
        } else {
            inner.resolveParameter(parameterContext, extensionContext)
        }
    }

    override fun afterEach(context: ExtensionContext) {
        val localWiremockResolver = localResolvers[context.uniqueId]
        if (localWiremockResolver != null) {
            localWiremockResolver.afterEach(context)
            localResolvers.remove(context.uniqueId)
        }
        inner.afterEach(context)
    }

    /**
     * Decorates a parameter also annotated with [WiremockResolver.Wiremock]
     */
    @Target(allowedTargets = [AnnotationTarget.VALUE_PARAMETER])
    @Retention(AnnotationRetention.RUNTIME)
    annotation class MethodScopedWiremockServer
}

  @Test
    fun `publishes to two Nexus repositories`(@MethodScopedWiremockServer @Wiremock otherServer: WireMockServer) {
    // omitted, rest of test
```

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (ryandens) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:50]: I'm unfamiliar with the `@Output` annotation - could you point me to some documentation? I don't see it in the `org.gradle.api.tasks` package and was under the impression that task outputs generally were associated with a file/directory

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt
Comment (ryandens) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt:280]: Agreed! When I first wrote this test, I didn't realize I would be adding this into the chain of execution of the publish tasks 😃

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (ryandens) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:74]: This makes sense, I agree that we don't want this to be printed out every time this task is executed. My first thought was to add a configuration property in the `GetStagingProfile` which determines whether or not the diagnostic data is printed out. That task property could be configured in the `NexusPublishPlugin`. However, I don't see a way of doing that without creating a new registration of the task with a different name to allow for the two different configurations to exist

In my opinion, that means the `GetStagingProfile` task is doing too much, as it trying to be a diagnostic tool and a first-class step in the publishing task chain. I think it would make sense to register a  `diagnosticGet${repository.name.capitalize()}StagingProfile` task for each repository that depends on the `get${repository.name.capitalize()}StagingProfile` task. The `GetStagingProfile` class is solely responsible for determining the staging profile id, either from configuration or from the HTTP API of the configured nexus repository. the `DiagnosticGetStagingProfile` consumes the staging profile ID and is instead focused on printing out the right message, depending on whether or not a `NexusRepository.stagingProfileId` is configured. This means that the `initialize` task can rely strictly upon the `GetStagingProfile` task but users still have access to a diagnostic message via the CLI if they deliberately execute the task.  This might look like:

```kotlin
/**
 * Diagnostic task for printing out the provided [stagingProfileId] in an informative message. The message
 * format differs based on whether or not the [NexusRepository.stagingProfileId] was configured explicitly
 * or not. If it was not, the [stagingProfileId] is presumed to have been obtained from a remote source,
 * like the [NexusRepository] HTTP API.
 */
@Suppress("UnstableApiUsage")
@Incubating
open class DiagnosticGetStagingProfile @Inject constructor(
    objects: ObjectFactory,
    extension: NexusPublishExtension,
    repository: NexusRepository
) : AbstractNexusStagingRepositoryTask(objects, extension, repository) {


    /**
     * This property is [Optional] because it is only required if [NexusRepository.stagingProfileId] is not provided
     */
    @Input
    @Optional
    val packageGroup = objects.property<String>().apply {
        set(extension.packageGroup)
    }

    @Input
    val stagingProfileId = project.objects.property(String::class)

    @TaskAction
    fun diagnosticGetStagingProfile() {
        val repository = repository.get()
        val configuredStagingProfileId = repository.stagingProfileId.orNull
        if (configuredStagingProfileId == null) {
            logger.lifecycle("Received staging profile id: '{}' for package '{}' for repository '{}'", stagingProfileId.get(), packageGroup.get(), repository.capitalizedName)
        } else {
            logger.lifecycle("Configured staging profile id: '{}' for repository '{}'", configuredStagingProfileId, repository.capitalizedName)
        }
    }
}
```

In `NexusPublishPlugin`, we would register the task and configure the properties and task dependencies for each repository

```kotlin
val diagnosticGetStagingProfile = rootProject.tasks.register<DiagnosticGetStagingProfile>("diagnosticGet${repository.name.capitalize()}StagingProfile")

getStagingProfileTask {
    description = "Gets a staging profile id for a given repository. If the staging profile id is configured, this task simply retrieves it from the extension that configured it. If the it is not, the staging profile id is retrieved from the NexusRepository via its HTTP API a packageGroup must be configured on the project that applied this plugin."
      group = PublishingPlugin.PUBLISH_TASK_GROUP
      stagingProfileIdOutput.set(project.layout.buildDirectory.file("publish-plugin/get${capitalizedName}StagingProfileOutput.txt"))
}
diagnosticGetStagingProfile {
    description = "Displays staging profile id for a given repository. This is a diagnostic task to get the value and put it into the NexusRepository configuration closure as stagingProfileId."
    group = PublishingPlugin.PUBLISH_TASK_GROUP
    stagingProfileId.set(getStagingProfileTask.map { it.stagingProfileIdOutput.get().asFile.readText() })
}

```

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (vlsi) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:53]: 1. `@Output` does not exist
2. By default, `Gradle` would consider the task `UP-TO-DATE` if all inputs and outputs are the same.

It looks like `OutputFile` is OK here.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt
Comment (vlsi) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt:82]: I'm not sure it is worth displaying the task in the common output for `./gradlew tasks`.
Don't get me wrong, but I believe very few end-users would call `./gradlew get... StagingProfile`.

Please keep `group=null` so the task does not appear in `gradlew tasks` output

File: src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt
Comment (vlsi) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt:195]: In practice, this dependency should be automatically set by Gradle when you configure `stagingProfileId.set(getStagingProfileTask.map...`

I guess it does not hurt, however, it is not needed either.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (vlsi) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:67]: I guess this should be factored into an `internal val` since the client instantiation is duplicated in all the tasks.

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt
Comment (marcphilipp) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt:238]: > Does JUnit Jupiter has an equivalent of `cleanup:` in Spock - to be always executed at the end of that particular test only?

It does not but it would be fairly easy to write an extension that collects `Closable` instances.

Comment (marcphilipp) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt:238]: > Does JUnit Jupiter has an equivalent of `cleanup:` in Spock - to be always executed at the end of that particular test only?

It does not but it would be fairly easy to write an extension that collects `Closable` instances.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (marcphilipp) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:53]: I don't think this task should ever be up-to-date since it depends on state on the server. Thus, if we remove the `@OutputFile` property and add an `@Internal val stagingProfileId: Property<String>` we don't need to write a file.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (marcphilipp) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:74]: I think we should at most log this on `info` level.

File: README.md
Comment (marcphilipp) [inline README.md:123]: I think we should name the task in a way that makes it clearer that it makes a call to a server, e.g. `retrieve${repository.name.capitalize()}StagingProfileId` or `find${repository.name.capitalize()}StagingProfileId`.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:53]: Sorry @ryandens for twisting the things around. I liked the idea with dependent tasks, however, the recent comments by Marc helped me realize that fetching the staging repository id from a server for diagnostic purpose is not the same as resolving staging repository id for the initialize tasks (and others). In the first case, server should be always asked (should ignore configured stagingProfileId), output should be printed on lifecycle level (people always want to see it) and it should be always not up-to-date. On the other side, for the initialize task is perfectly fine to get pre-configured value (if available), be quiet and play nicely with caching (assuming that `stagingProfileId` was set).

As a result, I think that we need `Retrieve<<Repo>>StagingProfile` task to implement the first case (always using  `client.findStagingProfileId()`). The second one could be covered as in the original version in `InitializeNexusStagingRepository`. As it is recommended to set `stagingProfileId` in the project configuration (for the performance reasons), I don't really see practical use case where someone would need to get that value from server on its own to reuse (with `stagingRepositoryId` the situation is completely different). WDYT?

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:74]: With my [last comment](https://github.com/gradle-nexus/publish-plugin/pull/62#discussion_r579826757) the whole case with different logging would disappear.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt:82]: Assuming its one-time per project (or even a group of projects) role, I agree that keeping it mentioned only in the documentation (in README) would be enough.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:67]: I believe, I proposed it in one of my earlier comments, but suggested to move it to a separate PR, after this one is merged to do not mix the things.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt
Comment (marcphilipp) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt:195]: Good point. We don't actually want that dependency when `stagingProfileId` is set explicitly. Thus, we should remove it.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:53]: @ryandens Are you ok with the proposed modifications (simplifications)?

Comment (ryandens) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt:53]: Yes, this seems reasonable! I'll take a stab at getting this all sorted out

File: src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt
Comment (ryandens) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt:82]: Definitely happy to defer to you all on this one, but for what it's worth the first thing I did when trying to figure out how to get the staging profile with this new plugin was running `./gradlew tasks` and checking the output for a similar task

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt
Comment (ryandens) [inline src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt:235]: this test isn't strictly necessary in order for us to feel confident about the changes that are now being made in this PR, but if you find it valuable I'm happy to leave it in.

Comment (szpak) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPluginTests.kt:235]: Yes, please.

**Review by szpak — APPROVED** [review]

File: src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt
Comment (marcphilipp) [inline src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt:31]: @szpak Not sure if we should use this annotation. WDYT?

**Review by marcphilipp — COMMENTED** [review]
Why don't we use the new task to retrieve the staging profile id for the initialize task?

File: src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt:31]: Personally, I like to use `@Incubating` for the new tasks/features to gather feedback and possibly have a clear way to fix spotted issues even in a non-backward compatible way. However, you don't like the "beta" state of that feature or just a Gradle annotation?

**Comment by szpak** [comment]
> Why don't we use the new task to retrieve the staging profile id for the initialize task?

@marcphilipp I asked the PR's author about that a comment earlier - see my [reasoning](https://github.com/gradle-nexus/publish-plugin/pull/62#discussion_r579826757). It simplifies the general construction and we eventually still use the common logic from the client.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt
Comment (marcphilipp) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt:31]: Ok, fine with me but we should have some common understanding when to do this. IIUC you'd like to do this for all new features. Is that correct?

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt
Comment (marcphilipp) [inline src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt:34]: Extensions should not have mutable state but store it in the `ExtensionContext.Store` but we/I can address that after this PR is merged.

**Review by marcphilipp — APPROVED** [review]

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt
Comment (ryandens) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt:34]: Ah, definitely agreed! This was easy enough for me to fix now 😃  - I'm a fan of Extension API so always enjoy the opportunity to work with it more

File: src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt
Comment (marcphilipp) [inline src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt:61]: There's no need to add the test method to the namespace since the extension context is already scoped to the method level.

Comment (ryandens) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt]: Thanks! I couldn't remember the rules around this so I went looking in the user guide and derived it from this example: https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks-before-after-execution  - but fixed!

Comment (marcphilipp) [reply src/compatTest/kotlin/io/github/gradlenexus/publishplugin/MethodScopeWiremockResolver.kt]: Good point! That example should indeed be simplified.

File: src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/RetrieveStagingProfile.kt:31]: It would be safe, but maybe it would be to defensive, as we - in general - do not expose API to others (at least, it is not the main goal of that extension). How does it look like in JUnit? Are you satisfied with the strategy adopted there?

File: src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt
Comment (szpak) [reply src/main/kotlin/io/github/gradlenexus/publishplugin/GetStagingProfile.kt]: Created #69.

**Comment by szpak** [comment]
Thanks @ryandens for your PR (and patience ;-) ). It will be a part of 1.1.0.

**Comment by ryandens** [comment]
Awesome, thanks for accepting my contribution and working through the feedback with me! Hoping this makes it easier for other folks to adopt this plugin 😄

**Pull request merged**
